pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/AbstractNode.java

```diff
@@ -425,22 +425,91 @@ public void remove() {
         // Detach current node of its parent, if any
         final Node parent = jjtGetParent();
         if (parent != null) {
-            parent.removeChildAtIndex(jjtGetChildIndex());
-            jjtSetParent(null);
+            parent.remove(jjtGetChildIndex());
         }
+    }
+
+    @Override
+    public void remove(final int index) {
+        if (0 > index || index >= jjtGetNumChildren()) {
+            return;
+        }
+
+        // Null child may have been caused due to an invalid insertion/addition
+        final Node oldChild = Objects.requireNonNull(children[index]);
 
-        // TODO [autofix]: Notify action for handling text edition
+        // Remove the child at the given index
+        children = ArrayUtils.remove(children, index);
+        // Update the remaining & left-shifted children indexes
+        for (int i = childIndex; i < jjtGetNumChildren(); i++) {
+            jjtGetChild(i).jjtSetChildIndex(i);
+        }
+
+        // Detach old child node of its parent, if any
+        oldChild.jjtSetParent(null);
+
+        // Finally, report the remove event
+        // removeChildEvent(this, oldChild, index); // TODO [autofix]
     }
 
     @Override
-    public void removeChildAtIndex(final int childIndex) {
-        if (0 <= childIndex && childIndex < jjtGetNumChildren()) {
-            // Remove the child at the given index
-            children = ArrayUtils.remove(children, childIndex);
-            // Update the remaining & left-shifted children indexes
-            for (int i = childIndex; i < jjtGetNumChildren(); i++) {
-                jjtGetChild(i).jjtSetChildIndex(i);
-            }
+    public int insert(final Node newChild, final int index) {
+        Objects.requireNonNull(newChild);
+        if (index < 0) {
+            return index;
+        }
+        final int numChildren = jjtGetNumChildren();
+        final int insertionIndex = index <= numChildren ? index : numChildren;
+        // Ensure that the given index position is empty
+        makeSpaceForNewChild(insertionIndex);
+        // Once shifted, the given index is empty. Let's add it as a new child
+        children[insertionIndex] = newChild;
+        newChild.jjtSetChildIndex(insertionIndex);
+        newChild.jjtSetParent(this);
+        // Finally, report the insert event
+        // insertChildEvent(this, newChild, insertionIndex); // TODO [autofix]
+        return insertionIndex;
```

Comment:
why are we duplicating all logic currently present in `jjtAddChild` on this same class?

Reply:
In fact, I think that we are duplicating only two lines of that method, which are:
```java
children[index] = child;
child.jjtSetChildIndex(index);
```

When calling `jjtAddChild`, the given child is inserted at the given index, without shifting any possible existing elements to the right (i.e., if there is a child at that index, it will perform an incomplete replacement - incomplete because it won't correctly unlink the old node being replaced).
However, when calling `insert`, the inner call to `makeSpaceForNewChild` ensures not only that the `children` array has enough space to insert the given new child at the given index, but also that all children from that index on are shifted to the right if there is an existing child at that index. This method also updates all children indexes if a shift is indeed performed.

Note that the `newChild.jjtSetParent(this);` statement in the `insert` method is not performed in the `jjtAddChild` method (which, in my opinion, should be done to fully add/link a child to its parent).

With all this, I consider that currently the logic is not being duplicated.

But please, let me know if you see a way to merge these two methods (`jjtAddChild` and `insert`), or if you see with good eyes that `jjtAddChild` just calls the `insert` method.

Reply:
True... still I think it will be very confusing on when to use which method... We need to be far more consistent on our APIs.